server/aesdsocket.c

```diff
@@ -91,33 +124,310 @@ void signal_handler(int signo)
+static void timer_init(int* fd, timer_t *timerid)
+{
+    struct sigevent sev;
+    struct itimerspec itimerspec;
+    int clock_id = CLOCK_MONOTONIC;
+    struct timespec start_time = {0};
+    int status;
+
+
+
+    //memset(&td,0,sizeof(struct thread_data));
+    memset(&sev,0,sizeof(struct sigevent));
+
+    sev.sigev_notify = SIGEV_THREAD;
+    sev.sigev_value.sival_ptr = fd;
+    sev.sigev_notify_function = timer_thread;
+
+    if(timer_create(clock_id, &sev, timerid) != 0)
+    {
+        LOG_ERROR("Error creating timer\n");
+        return;
+    }
+    
+    status = clock_gettime(clock_id, &start_time);
+    if(status != 0)
+    {
+        perror("Error getting clock\n");
+        return;
+    }
+
+    itimerspec.it_interval.tv_sec = 10;
+    itimerspec.it_interval.tv_nsec = 1000000;//itimerspec.it_value.tv_nsec;
+
+     
+    timespec_add(&itimerspec.it_value,&start_time,&itimerspec.it_interval);
+    
+    status = timer_settime(*timerid, TIMER_ABSTIME, &itimerspec, NULL );
+    if(status != 0)
+    {
+        LOG_ERROR("Error getting clock\n");
+        return;
+    }
+    
+}
+
+
+/******************************************************************************
+ * socket communication on every accepted connection
+ * ****************************************************************************/
+void *socket_communication(void *threadp)
+{
+
     long nread, nwrite, nsent, nreceived;
-    struct sockaddr_in server_addr, new_addr;
-    struct sockaddr_in *pV4Addr = (struct sockaddr_in *)&new_addr;
-    struct in_addr client_addr = pV4Addr->sin_addr;
+    char *heap_buffer_for_write = NULL;
+    char *heap_buffer_for_read = NULL;
+    long filled_up_buffer_size = 0;
     char buf[BUF_SIZE];
     long total_buffer_size_for_write_into_file = BUF_SIZE;
     long total_buffer_size_for_read_from_file = BUF_SIZE;
-    long filled_up_buffer_size = 0;
-    socklen_t addr_size;
-    bool daemon_mode = false;
-    char str[INET_ADDRSTRLEN];
     char *newline_char_ptr;
     long newline_char_index = 0;
     long prev_newline_char_index = 0;
     long end_of_file_index = 0;
+    //char str[INET_ADDRSTRLEN];
 
-    char *heap_buffer_for_write = NULL;
-    char *heap_buffer_for_read = NULL;
-    int return_status = 0;
+    threadParams_t *thread_params = (threadParams_t *)threadp;
+
+    
+    heap_buffer_for_write = (char *)malloc(sizeof(char) * BUF_SIZE);
+    if (heap_buffer_for_write == NULL)
+    {
+        LOG_ERROR("could not malloc");
+        server_socket_state = STATE_EXIT;
+    }
+    server_socket_state = STATE_RECEIVE_FROM_CLIENT;
+
+    
+    while (1)
+    {
+        switch (server_socket_state)
+        {
+        case STATE_RECEIVE_FROM_CLIENT:
+            LOG_DBG("STATE_RECEIVE_FROM_CLIENT\n");
+            do
+            {
+                nreceived = recv(thread_params->sockfd_client, buf, sizeof(buf), 0);
+                if (!nreceived || (strchr(buf, '\n') != NULL))
```

Comment:
@rajatchaple I think you could have a problem here if you read `sizeof(buf)` because you might not have a string terminator, so this `strchr` would potentially read uninitialized memory which may or may not be 0.  You should make sure your buffer is long enough to hold a string terminator and then make sure you write one after the last byte received.